### PR TITLE
Add `ModelTesterMixin`, UI improvement 

### PR DESCRIPTION
**Pull request by LysandreJik**
This PR aims to add the mixin `ModelTesterMixin` to LXMERT, to ensure that it behaves correctly. In doing so there has been a few bugfixes, but also a slight refactor of LXMERT's UI. This affects only user-facing method of user-facing models (`LxmertModel`, `LxmertForPretraining`, `LxmertForQuestionAnswering`). 

Some things to note: 
- the UI is extremely important. The UI need to align with the rest of the library or users will be lost and either open issues on a regular basis to understand, or simply not use the model at all.
- The UI is the most important component of the model, as it cannot be easily changed over time. Once the UI is set, introducing any changes to it results in breaking changes which is a tremendous pain for users. The internals of the models may change over time, however, as long as the resulting behavior doesn't change.

The noteworthy changes are detailed in comments below for an easier review.

Only PyTorch for now; if you agree with these changes I'll do TensorFlow as well.

### CODE REVIEW

File: docs/source/model_doc/lxmert.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
+LXMERT
+----------------------------------------------------
+
+Overview
+~~~~~~~~~~~~~~~~~~~~~
+
+
+
```
Comment (LysandreJik) [inline docs/source/model_doc/lxmert.rst:8]: This adds LXMERT to the documentation. I've left the `Overview` blank for now.

File: src/transformers/configuration_lxmert.py [hunk 2]
```diff
@@ -176,3 +176,8 @@ def __init__(
         self.visual_feat_loss = visual_feat_loss
         self.output_hidden_states = output_hidden_states
         self.output_attentions = self.output_attentions
+        self.num_hidden_layers = {
+            "vision": r_layers,
+            "cross_encoder": x_layers,
+            "language": l_layers
+        }
```
Comment (LysandreJik) [inline src/transformers/configuration_lxmert.py:183]: This is very different to what we usually do, but as it's the first multi-transformer architecture we can have a bit more freedom. I think this approach makes sense, but we'll have to discuss with other team members before the final merge in `master` if we decide to merge this.

File: src/transformers/modeling_lxmert.py [hunk 3]
```diff
@@ -27,13 +27,16 @@
...
+from .file_utils import ModelOutput, add_start_docstrings, add_start_docstrings_to_callable, add_code_sample_docstrings, \
+    replace_return_docstrings
 from .modeling_utils import PreTrainedModel
 
 
 logger = logging.getLogger(__name__)
 
+_CONFIG_FOR_DOC = "LxmertConfig"
+_TOKENIZER_FOR_DOC = "LxmertTokenizer"
+
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:39]: Our documentation format has slightly changed, I've updated it here.

File: src/transformers/modeling_lxmert.py [hunk 4]
```diff
@@ -46,6 +49,96 @@ def __init__(self):
     def forward(self, x):
         return gelu(x)
 
+@dataclass
+class LxmertModelOutput(ModelOutput):
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:53]: Moved the model outputs in this file, as they're model-specific. Renamed and re-ordered the parameters.

File: src/transformers/modeling_lxmert.py [hunk 5]
```diff
@@ -46,6 +49,96 @@ def __init__(self):
...
+            Tuple of :obj:`torch.FloatTensor` (one for each layer) of shape
+            :obj:`(batch_size, num_heads, sequence_length, sequence_length)`.
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
+        cross_encoder_attentions (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
+            Tuple of :obj:`torch.FloatTensor` (one for each layer) of shape
+            :obj:`(batch_size, num_heads, sequence_length, sequence_length)`.
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
+    """
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:89]: Several things we have to respect with these model outputs:

- There should be absolutely no different behavior between using the traditional tuple output and this model output, if using `.to_tuple`. This means that:
```py
model(**inputs) == model(**inputs, return_dict=True).to_tuple()
```
We're adding a test to the common tests about this [here](https://github.com/huggingface/transformers/pull/6445).
- Model outputs should therefore have the same order as the regular tuple outputs.
- I've re-ordered it as such: CATEGORY_1[language, vision, misc], CATEGORY_2[language, vision, misc], etc.
- The documented args should have the same order as the args defined in the model output.

File: src/transformers/modeling_lxmert.py [hunk 6]
```diff
@@ -62,39 +155,39 @@ class LxmertForPretrainingOutput(ModelOutput):
...
-    hidden_states_l: Optional[Tuple[torch.FloatTensor]] = None
-    hidden_states_v: Optional[Tuple[torch.FloatTensor]] = None
-    attentions_v_encoder: Optional[Tuple[torch.FloatTensor]] = None
-    attentions_l_encoder: Optional[Tuple[torch.FloatTensor]] = None
-    attentions_x_encoder: Optional[Tuple[torch.FloatTensor]] = None
+    language_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+    vision_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+    language_attentions: Optional[Tuple[torch.FloatTensor]] = None
+    vision_attentions: Optional[Tuple[torch.FloatTensor]] = None
+    cross_encoder_attentions: Optional[Tuple[torch.FloatTensor]] = None
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:190]: Here we can (and should) take full advantage of the model output to have the most explicit key names as possible. No need for abbreviations.

File: src/transformers/modeling_lxmert.py [hunk 7]
```diff
@@ -189,18 +282,27 @@ def __init__(self, config):
...
+        if input_ids is not None:
+            input_shape = input_ids.size()
+            device = input_ids.device
+        else:
+            input_shape = inputs_embeds.size()[:-1]
+            device = inputs_embeds.device
+        seq_length = input_shape[1]
+
+        position_ids = torch.arange(seq_length, dtype=torch.long, device=device)
+        position_ids = position_ids.unsqueeze(0).expand(input_shape)
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:295]: These modifications and the ones below are done in order to accept `inputs_embeds` for the language transformer.

File: src/transformers/modeling_lxmert.py [hunk 8]
```diff
@@ -343,7 +445,7 @@ def __init__(self, config):
         self.output = LxmertOutput(config)
 
     def forward(self, hidden_states, attention_mask=None, head_mask=None, output_attentions=False):
-        outputs = self.attention(hidden_states, hidden_states, attention_mask, output_attentions=output_attentions)
+        outputs = self.attention(hidden_states, attention_mask, output_attentions=output_attentions)
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:448]: This is duplicated in the attention which is already doing 
```py
output = self.self(input_tensor, input_tensor, attention_mask, output_attentions=output_attentions)
```

File: src/transformers/modeling_lxmert.py [hunk 9]
```diff
@@ -400,15 +502,15 @@ def forward(self, lang_feats, lang_attention_mask, visn_feats, visn_attention_ma
         visn_att_output = (visn_feats,)
 
         lang_att_output, visn_att_output = self.cross_att(
-            lang_att_output[0], lang_attention_mask, visn_att_output[0], visn_attention_mask,
+            lang_att_output[0], lang_attention_mask, visn_att_output[0], visn_attention_mask, output_x_attentions=output_attentions
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:505]: Was not returning the attentions because the flag wasn't passed.

File: src/transformers/modeling_lxmert.py [hunk 10]
```diff
@@ -400,15 +502,15 @@ def forward(self, lang_feats, lang_attention_mask, visn_feats, visn_attention_ma
...
         )
         attention_probs = lang_att_output[1:]
         lang_att_output, visn_att_output = self.self_att(
             lang_att_output[0], lang_attention_mask, visn_att_output[0], visn_attention_mask
         )
 
         lang_output, visn_output = self.output_fc(lang_att_output, visn_att_output)
-        return (lang_output, visn_output) + (attention_probs,) if output_attentions else (lang_output, visn_output)
+        return (lang_output, visn_output) + (attention_probs[0],) if output_attentions else (lang_output, visn_output)
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:513]: Here `attention_probs = lang_att_output[1:]`, with `len(lang_att_output)` in [0, 1]. 

If 1, then `lang_att_output[1:]` returns a tuple containing a single variable, and we then concatenate that tuple to the return tuple, resulting in a tuple containing a tuple of a tensor, instead of a tuple containing only tensors.

File: src/transformers/modeling_lxmert.py [hunk 11]
```diff
@@ -473,27 +575,27 @@ def forward(
...
             v_outputs = layer_module(visn_feats, visn_attention_mask, output_attentions=output_attentions)
             visn_feats = v_outputs[0]
-            if all_v_attentions:
+            all_v_hidden_states = all_v_hidden_states + (visn_feats,)
+            if all_v_attentions is not None:
                 all_v_attentions = all_v_attentions + (v_outputs[1],)
 
         # Run cross-modality layers
-        all_v_hidden_states = all_v_hidden_states + (visn_feats,)
-        all_l_hidden_states = all_l_hidden_states + (lang_feats,)
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:488]: Attentions were not correctly added.

File: src/transformers/modeling_lxmert.py [hunk 12]
```diff
@@ -473,27 +575,27 @@ def forward(
...
-            if all_l_attentions:
+            all_l_hidden_states = all_l_hidden_states + (lang_feats,)
+            if all_l_attentions is not None:
                 all_l_attentions = all_l_attentions + (l_outputs[1],)
 
         # Run relational layers
         for layer_module in self.r_layers:
             v_outputs = layer_module(visn_feats, visn_attention_mask, output_attentions=output_attentions)
             visn_feats = v_outputs[0]
-            if all_v_attentions:
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:483]: `bool(()) == False`, need to enforce that it is not `None`.

File: src/transformers/modeling_lxmert.py [hunk 13]
```diff
@@ -660,45 +762,46 @@ def _init_weights(self, module):
             :func:`transformers.PreTrainedTokenizer.__call__` for details.
 
             `What are input IDs? <../glossary.html#input-ids>`__
+        visual_feats: (:obj:`Tuple[torch.FloatTensor, torch.FloatTensor]`):
+            the first item in the tuple represents the actual visual features
+            (ROI pooled object features from bounding boxes using a faster-RCNN model) of shape ՝(batch_size, num_visual_features, visual_feat_dim)՝
+            and where the second item represents the normalized bounding boxes on a scale of 0~1  of shape ՝(batch_size, 4)՝.
+            These are currently not provided by the transformers library
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:769]: Here as well, documentation order must respect argument order.

File: src/transformers/modeling_lxmert.py [hunk 14]
```diff
@@ -714,28 +817,49 @@ def get_input_embeddings(self):
     def set_input_embeddings(self, new_embeddings):
         self.embeddings.word_embeddings = new_embeddings
 
-    def _prune_heads(self, *args, **kwargs):
-        pass
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:718]: This would potentially prevent from having a `NotImplementedError`, instead silently passing.

File: src/transformers/modeling_lxmert.py [hunk 15]
```diff
@@ -761,47 +885,55 @@ def forward(
...
-            attentions_v_encoder=v_outputs[1:] if output_attentions else None,
-            attentions_x_encoder=x_encoder_outputs[2:] if output_attentions else None,
+            pooled_output=pooled_output,
+            language_output=lang_output,
+            vision_output=visn_output,
+            language_hidden_states=all_lang_hidden_states if output_hidden_states else None,
+            vision_hidden_states=all_visn_hidden_states if output_hidden_states else None,
+            language_attentions=all_l_attentions if output_attentions else None,
+            vision_attentions=all_v_attentions if output_attentions else None,
+            cross_encoder_attentions=all_x_attentions if output_attentions else None,
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:931]: This is very important. It updates to more coherent returns, with a flat list of returns instead of the previous tuple of tuples.

The tuple returns must have the same value as the model outputs if cast to a tuple.

Comment (LysandreJik) [reply src/transformers/modeling_lxmert.py:931]: Please note that this will potentially break your existing code bases if relying on this code.

File: src/transformers/modeling_lxmert.py [hunk 16]
```diff
@@ -923,79 +1055,71 @@ def _get_resized_qa_labels(self, cur_qa_logit_layer, num_labels):
...
-            and where the second item represents the normalized bounding boxes on a scale of 0~1  of shape ՝(batch_size, 4)՝.
-            These are currently not provided by the transformers library
-        obj_labels: (``Dict[Str: Tuple[Torch.FloatTensor, Torch.FloatTensor]]``, `optional`, defaults to :obj: `None`):
-            each key is named after each one of the visual losses and each element of the tuple is of the shape
-            ``(batch_size, num_features)`` and ``(batch_size, num_features, visual_feature_dim)``
-            for each the label id and the label score respectively
         ans: (``Torch.Tensor`` of shape ``(batch_size)``, `optional`, defaults to :obj: `None`):
             a one hot representation hof the correct answer `optional`
-        kwargs: (:obj:`Dict[str, any]`, optional, defaults to `{}`):
-            Used to hide legacy arguments that have been deprecated.
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:975]: Using kwargs results in more errors done in tests, and since it's the introduction of the model there's no need for legacy arguments management.

File: src/transformers/modeling_lxmert.py [hunk 17]
```diff
@@ -923,79 +1055,71 @@ def _get_resized_qa_labels(self, cur_qa_logit_layer, num_labels):
...
+        lang_output, visn_output, pooled_output = lxmert_output[0], lxmert_output[1], lxmert_output[2]
+
         lang_prediction_scores, cross_relationship_score = self.cls(lang_output, pooled_output)
         if self.task_qa:
             answer_score = self.answer_head(pooled_output)
         else:
             answer_score = pooled_output[0][0]
 
-        total_loss = 0.0
+        total_loss = torch.tensor(0.0, device=device)
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:1122]: We deal in tensor or tuple of tensor outputs only.

File: src/transformers/modeling_lxmert.py [hunk 18]
```diff
@@ -1069,7 +1187,7 @@ def __init__(self, config):
         self.init_weights()
 
         # Loss function
-        self.loss = CrossEntropyLoss(ignore_index=-100)
+        self.loss = CrossEntropyLoss()
```
Comment (LysandreJik) [inline src/transformers/modeling_lxmert.py:1190]: The ignore_index is -100 by default

File: tests/test_modeling_lxmert.py [hunk 19]
```diff
@@ -246,11 +251,11 @@ def create_and_check_lxmert_model(
...
         result = model(input_ids, (visual_feats, bounding_boxes), return_dict=True)
 
-        self.parent.assertEqual(result.last_hidden_state_l.shape, (self.batch_size, self.seq_length, self.hidden_size))
+        self.parent.assertEqual(result.language_output.shape, (self.batch_size, self.seq_length, self.hidden_size))
         self.parent.assertEqual(
-            result.last_hidden_state_v.shape, (self.batch_size, self.num_visual_features, self.hidden_size)
+            result.vision_output.shape, (self.batch_size, self.num_visual_features, self.hidden_size)
         )
-        self.parent.assertEqual(result.pooled_output_x_encoder.shape, (self.batch_size, self.hidden_size))
+        self.parent.assertEqual(result.pooled_output.shape, (self.batch_size, self.hidden_size))
```
Comment (LysandreJik) [inline tests/test_modeling_lxmert.py:258]: I hope you'll agree with me that this output is cleaner and simpler to understand for users :)

File: tests/test_modeling_lxmert.py [hunk 20]
```diff
@@ -518,10 +513,14 @@ def prepare_config_and_inputs_for_common(self):
 
 
 @require_torch
-class LxmertModelTest(unittest.TestCase):
+class LxmertModelTest(ModelTesterMixin, unittest.TestCase):
```
Comment (LysandreJik) [inline tests/test_modeling_lxmert.py:516]: boom!

File: tests/test_modeling_lxmert.py [hunk 21]
```diff
@@ -550,3 +549,139 @@ def test_model_from_pretrained(self):
         for model_name in LXMERT_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
             model = LxmertModel.from_pretrained(model_name)
             self.assertIsNotNone(model)
+
+    def test_attention_outputs(self):
```
Comment (LysandreJik) [inline tests/test_modeling_lxmert.py:553]: Needed to re-implement that test here, as well as the hidden states test since the behavior is different to other single-transformer models.

File: src/transformers/modeling_lxmert.py (hunk 12 above)
Comment (eltoto1219) [reply src/transformers/modeling_lxmert.py:483]: Ah so that is why it was "None" before, makes a lot of sense. I can definetely fix this
